consumer/src/main/java/com/flipkart/varadhi/consumer/GroupedMessageSrc.java

```diff
@@ -0,0 +1,178 @@
+package com.flipkart.varadhi.consumer;
+
+import com.flipkart.varadhi.entities.Message;
+import com.flipkart.varadhi.entities.Offset;
+import com.flipkart.varadhi.spi.services.Consumer;
+import com.flipkart.varadhi.spi.services.PolledMessage;
+import com.flipkart.varadhi.spi.services.PolledMessages;
+import lombok.AllArgsConstructor;
+import lombok.RequiredArgsConstructor;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.mutable.MutableBoolean;
+
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Message source that maintains ordering among messages of the same groupId.
+ */
+@RequiredArgsConstructor
+public class GroupedMessageSrc<O extends Offset> implements MessageSrc {
+
+    private final ConcurrentHashMap<String, GroupTracker> allGroupedMessages = new ConcurrentHashMap<>();
+    private final ConcurrentLinkedDeque<String> freeGroups = new ConcurrentLinkedDeque<>();
+
+    /**
+     * Maintains the count of total messages read from the consumer so far.
+     * Required for watermark checks, for when this value runs low we can fetch more messages from the consumer.
+     * Counter gets decremented when the message is committed/consumed.
+     */
+    private final AtomicLong totalInFlightMessages = new AtomicLong(0);
+
+    // Used for watermark checks against the totalInFlightMessages. Will be driven via consumer configuration.
+    private final long maxInFlightMessages = 100; // todo(aayush): make configurable
+
+    private final Consumer<O> consumer;
+
+    /**
+     * Attempt to fill the message array with one message from each group.
+     * Subsequent messages from a group are not fetched until the previous message is consumed.
+     *
+     * @param messages Array of message trackers to populate.
+     *
+     * @return CompletableFuture that completes when the messages are fetched.
+     */
+    @Override
+    public CompletableFuture<Integer> nextMessages(MessageTracker[] messages) {
+        if (hasMaxInFlightMessages()) {
+            return replenishAvailableGroups().thenApply(v -> nextMessagesInternal(messages));
+        }
+        return CompletableFuture.supplyAsync(() -> nextMessagesInternal(messages));
```

Comment:
It does not have to be CompletableFuture.supplyAsync

You can use CompletableFuture.completedFuture(), to imediately return with result. this supplyAsync introduces a unnecessary jump to another thread.